### PR TITLE
Using distribution reference pkg (pin image by digest)

### PR DESCRIPTION
**Pull request by nishanttotla**
This PR is a follow up to #28173.

- The current code that extracts a digest from the service image in `docker service create` and `docker service update` uses `ParseNamed` from the `docker/docker/reference` package. One problem this causes is that tags are dropped if both tag and digest is provided. For instance , 

```
docker service create alpine:latest@sha256:1354db23ff5478120c980eca1611a51c9f2b88b61f24283ee8200bf9a54f2e5c top
```
will create a service, but the service image will be listed as `alpine@sha256:1354db23ff5478120c980eca1611a51c9f2b88b61f24283ee8200bf9a54f2e5c` instead of `alpine:latest@sha256:1354db23ff5478120c980eca1611a51c9f2b88b61f24283ee8200bf9a54f2e5c`. Using `ParseNamed` from `docker/distribution/reference` package fixes that.

- The second problem this PR fixes is that it gets rid of string manipulation, and uses the `WithDigest` function from `docker/distribution/reference` to cleanly construct the image string.

- This PR also updates the vendored version for `docker/distribution` to include https://github.com/docker/distribution/pull/2062 and https://github.com/docker/distribution/pull/2044

### CODE REVIEW

File: daemon/cluster/cluster.go [hunk 1]
```diff
@@ -1031,15 +1037,11 @@ func (c *Cluster) imageWithDigestString(ctx context.Context, image string, authC
...
-		// TODO(nishanttotla): Currently, the service would lose the tag while calling WithDigest
-		// To prevent this, we create the image string manually, which is a bad idea in general
-		// This will be fixed when https://github.com/docker/distribution/pull/2044 is vendored
-		// namedDigestedRef, err := reference.WithDigest(ref, dscrptr.Digest)
-		// if err != nil {
-		// 	return "", err
-		// }
-		// return namedDigestedRef.String(), nil
-		return image + "@" + dscrptr.Digest.String(), nil
+		namedDigestedRef, err := distreference.WithDigest(distreference.EnsureTagged(ref), digest.Digest(dscrptr.Digest.String()))
```
Comment (aaronlehmann) [inline daemon/cluster/cluster.go]: Why convert `descrptr.Digest` to a string and back?

Comment (nishanttotla) [inline daemon/cluster/cluster.go]: @aaronlehmann you're right, this isn't needed. At first I assumed `descrptr.Digest` was a `docker/docker/distribution` object.

**Comment by aaronlehmann** [comment]
Before anyone asks, yes, it's highly unfortunate that this involves parsing the reference twice. @dmcgowan was working on unifying the two reference packages but this didn't make it into 1.13. I think the solution proposed here is better than the string manipulation we were doing before. As the reference packages converge, this will become cleaner.

Design LGTM

**Comment by nishanttotla** [comment]
I don't think the test failure is related. Is this a flaky test from another PR?

**Comment by dmcgowan** [comment]
LGTM

To keep with the spirit of the reference package, this function should evolve to take the image in as a `reference.Named` and return a `reference.Canonical`. We can do that after resolving the fork.

**Comment by nishanttotla** [comment]
@dmcgowan that would be great (for this and similar functions). Adding a TODO item to keep track of that.

**Comment by nishanttotla** [comment]
All tests seem to have passed this time.

cc @aaronlehmann @tonistiigi

**Comment by aaronlehmann** [comment]
LGTM

**Comment by tonistiigi** [comment]
LGTM

**Pull request merged**